### ISSUE TITLE
There are many problems in the code that are inconsistent with the description in the paper.

### ISSUE THREAD

**Issue by WentaoXie1998** [zhengxin-plog/MD-CT-DTA#2]
Key mismatches (paper description/figures vs public code)
2.1 CTN block (protein encoder)
Status: partially fixed; several inconsistencies remain.
In the original public implementation, the CTN block was implemented purely as a multi-branch 1D CNN module:
Embedding → three parallel StackCNN branches (with 1/2/3 Conv1d layers respectively) → AdaptiveMaxPool1d(1) → concatenation → a linear fusion layer;
The block returned (x, layer1, layer2, layer3), where layer1/2/3 were branch-wise pooled vectors of shape [B, 96], and x was the concatenated and linearly fused vector of shape [B, 96];
No MHSA (multi-head self-attention), no MLP, no LayerNorm, and no residual connections were present, which conflicted with the CTN description in the paper.
In the revised model.py, the CTN block has been extended as follows:
A new MHSA module (multi-head self-attention) and a small MLP are added inside CTNBlock;
In forward, the code performs:
1.Embedding and permutation: x = embed(seq).permute(0,2,1).
2.Three parallel StackCNN branches to produce layer1, layer2, and layer3 (each [B,96]).
3.Concatenation of the three branch outputs followed by self.linear to obtain a fused vector x of shape [B,96].
4.Application of MHSA and MLP with residual connections:
o1 = MHSA(x) + x
o2 = linear1(o1)
o3 = MLP(o2) + o1
However, the return value of the block remains return x, layer1, layer2, layer3, so the downstream model still receives the pre-MHSA fused vector x, not the fully updated output (o2 or o3).
There are two main remaining problems:
Attention is applied across the batch, not along the sequence.
	The MHSA module takes an input of shape [L, D] and treats the first dimension as the “sequence length” (L). In the CTN block, x passed to MHSA is of shape [B, 96], where B is the batch size. This means the self-attention is computed between different samples in the same mini-batch, rather than across positions within a single protein sequence, which does not match the usual Transformer-style MHSA described in the paper.
The CTN output used by the rest of the model ignores the MHSA+MLP refinement.
	Because the block returns x (before MHSA/MLP) rather than o2/o3, the newly added Transformer components do not actually affect the features that flow into the interaction module and final predictor. Conceptually the block is closer to the paper than before, but functionally the model still behaves as a pure multi-branch CNN encoder.
In addition, the convolutions remain nn.Conv1d(kernel_size=3), i.e. 1D kernels, while the paper’s figure still labels them as “3×3 Conv”. If the intended design is 1D convolution, the figure notation continues to be somewhat misleading; if a 2D kernel was intended, the implementation is not aligned with that description.
2.2 MGDC / GDC (drug encoder, “diffusion convolution”)
Status: partially fixed; diffusion is computed but not actually used as the node representation.
In the earlier implementation, the GDC module was essentially a two-layer GCN:
GDCconvLayerIn = GCNConv(in_channels, hidden_channels)
GDCconvLayerOut = GCNConv(hidden_channels, out_channels)
A Node-level batch norm and ReLU activation were applied.
The attributes diffusion_iterations and diffusion_weight were defined but never used in forward(), so the claimed “graph diffusion convolution” did not appear in the actual computation.
In the revised code, the GDC module is rewritten:
It still starts with two GCNConv layers plus NodeLevelBatchNorm and ReLU:
data.x = F.relu(self.norm(self.conv2(self.conv1(x, edge_index), edge_index)))
After that, it constructs a dense adjacency matrix adj_t from edge_index, adds self-loops, and applies an iterative update:
adj_t_{t+1} = diffusion_weight * adj_t_t @ adj_t_t + (1 - diffusion_weight) * I
Repeated for diffusion_iterations steps.
The resulting adj_t is row-normalized.
It then computes a Laplacian-like matrix and a scalar regularization term:
laplacian = I - adj_t
reg_term = trace(X^T laplacian)
Finally, it forms x = adj_t @ x + reg_term.
However, despite computing this diffused x, the method never assigns it back to data.x. The only assignment to data.x is the earlier GCNConv output. As a result:
Downstream GraphDenseNet blocks still operate on the GCNConv output only, and the computed diffusion result is effectively unused.
The scalar reg_term is added directly to x, which broadcasts the same scalar to all nodes and all feature channels; this is unusual for a “diffusion convolution” and is not clearly motivated in the paper.
Therefore, while the revised code now includes diffusion-related operations and parameters in the forward pass, the effective behavior from the graph encoder’s perspective remains that of a standard two-layer GCN, not the diffusion convolution described in the article.
	
2.3 MGDC block staging (multi-scale blocks and pooling)
Status: not fixed; the implementation is still parallel multi-branch GraphDenseNet, not the staged MGDC→Maxpool pipeline shown in the figure.
The paper’s figure presents a staged multi-scale pipeline on the drug side:
MGDC block (n=1) → Maxpool → MGDC block (n=2) → Maxpool → MGDC block (n=3) → Maxpool → Readout.
In both the original and the revised code, the actual implementation follows a different pattern:
GraphDenseNet is defined as a stack of DenseBlock modules with GDC transitions, and at the end a global mean pooling (global_mean_pool) followed by a linear classifier.
